### PR TITLE
Fix ledger replicated failed blocks bookie decommission process

### PR DESCRIPTION
**Pull request by hangc0276**
### Motivation
When I decommission one bookie (bk3), one ledger replicate failed and blocked decommission process.

This is the auto-recovery log:
```
2023-03-29T06:29:22,642+0000 [ReplicationWorker] ERROR org.apache.bookkeeper.client.LedgerHandle - ReadEntries exception on ledgerId:904368 firstEntry:14 lastEntry:14 lastAddConfirmed:13
2023-03-29T06:29:22,642+0000 [ReplicationWorker] ERROR org.apache.bookkeeper.replication.ReplicationWorker - Received error: -1 while trying to read entry: 14 of ledger: 904368 in ReplicationWorker
2023-03-29T06:29:22,642+0000 [ReplicationWorker] ERROR org.apache.bookkeeper.replication.ReplicationWorker - Failed to read faulty entries, so giving up replicating ledgerFragment Fragment(LedgerID: 904368, FirstEntryID: 0[0], LastKnownEntryID: 14[14], Host: [betausc1-bk-10.betausc1-bk-headless.o-vaxkx.svc.cluster.local:3181], Closed: true)
2023-03-29T06:29:22,644+0000 [ReplicationWorker] ERROR org.apache.bookkeeper.replication.ReplicationWorker - ReplicationWorker failed to replicate Ledger : 904368 for 6 number of times, so deferring the ledger lock release by 300000 msecs
```
The ledger's metadata:
```
ledgerID: 904368
2023-03-29T06:47:56,511+0000 [main] INFO  org.apache.bookkeeper.tools.cli.commands.
client.LedgerMetaDataCommand - LedgerMetadata{formatVersion=3, ensembleSize=3, writeQuorumSize=3, 
ackQuorumSize=2, state=OPEN, digestType=CRC32C, password=base64:, 
ensembles={0=[bk1:3181, bk2:3181, bk3:3181], 15=[bk1:3181, bk2:3181, bk4:3181]},...}
```

The ledger (904368) has two ensembles, `ensembles={0=[bk1:3181, bk2:3181, bk3:3181], 15=[bk1:3181, bk2:3181, bk4:3181]}`. However, the replication worker got the ledger's LAC is 13, but it got the replication fragment entry range is [0, 14]. When reading entry 14, it failed.

### One question
**Why the ensembles created a new ensemble starting with entryId = 15, but the ledger's lastAddConfirm is 13.**

This question is related to two parts, one is how the new ensemble was created and the other is how the lastAddConfirm was generated.

#### 1. How the new ensemble was created
The ensemble change is controlled on the bookie client side. 

When one entry is ready to send to the bookie server, the bookie client will check whether need to do the ensemble change.
https://github.com/apache/bookkeeper/blob/912896deb2e748389e15e74c37539b2ff36302c7/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingAddOp.java#L254

For the above case, when writing entry 15, one bookie is lost, it will trigger the ensemble change and generate the new ensemble: 15=[bk1:3181, bk2:3181, bk4:3181]. However, entry 15 write failed, such as timeout or bookie server rejected the write.

For now, entry 14 is written succeed.

#### 2. How the lastAddConfirm was generated
Due to the ledger being in the `OPEN` state, the ledger handle will send a readLAC request according to the last ensemble to get the ledger's lastAddConfirm. 

For the above case, the readLAC request will send to bk1, bk2, and bk4.

For the `V2` protocol (Pulsar uses the V2 protocol to interact with the BookKeeper cluster), the bookie client put the lastAddConfirm EntryId in the next Entry's metadata. 
https://github.com/apache/bookkeeper/blob/df4492012cc03682534cbc8dd68dd81163b0c947/bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/checksum/DigestManager.java#L134

When we use the `V2` protocol to open an `OPEN` state ledger to read, it will send a readLastAddConfirm request to the bookie server, and the bookie server gets the last entry of this ledger and return to the client.
https://github.com/apache/bookkeeper/blob/df4492012cc03682534cbc8dd68dd81163b0c947/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ReadLastConfirmedOp.java#L108

However, the bookie client will parse the response entry and get the lastAddConfirm from the entry's metadata. Due to the entry just recording the previous EntryId as the lastAddConfirm, the LedgerHandle got the lastAddConfirm will be the penultimate EntryId of the ledger.

For the above case, the bk1 holds the max entry 14, bk2 holds the max entry 14, and bk4 returns NoSuchEntryException, LedgerHandle gets lastAddConfirm will be `14  - 1 = 13`, not 14.

When the replicator tries to recover the first ensemble 0=[bk1:3181, bk2:3181, bk3:3181] with entry range [0, 14],  reading entry 14 will throw a ReadEntryException due to the lastAddConfirm is 13.
https://github.com/apache/bookkeeper/blob/df4492012cc03682534cbc8dd68dd81163b0c947/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java#L685-L690

### Solution
When encountered that case that 
- The ledger is `OPEN`
- The ledger has multiple ensembles
- The ledger's last ensemble doesn't have any entries, which means `lastAddConfirm < last ensemble key - 1`

We should treat the penultimate segment/ensemble of the ledger as an `OPEN` state instead of a closed state.
https://github.com/apache/bookkeeper/blob/df4492012cc03682534cbc8dd68dd81163b0c947/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragment.java#L56-L57

After we treat the segment/ensemble as `OPEN` state, the replicator will close the ledger first and replicate it.

### CODE REVIEW

**Review by eolivelli — COMMENTED** [review]
I wonder of we risk to lose one entry in case of AQ < WQ.

I have to think more

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookieWriteLedgerTest.java [hunk 1]
```diff
@@ -1454,6 +1467,67 @@ public void testLedgerCreateByteBufRefCnt() throws Exception {
...
+        }
+
+        List<BookieId> ensemble = lh.getLedgerMetadata().getAllEnsembles().entrySet().iterator().next().getValue();
+        assertEquals(1, lh.getLedgerMetadata().getAllEnsembles().size());
+        killBookie(ensemble.get(1));
+
+
+        try {
+            lh.ensembleChangeLoop(ensemble, Collections.singletonMap(1, ensemble.get(1)));
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (hangc0276) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookieWriteLedgerTest.java:1486]: done

**Comment by horizonzy** [comment]
> How the lastAddConfirm was generated
> Due to the ledger being in the OPEN state, the ledger handle will send a readLAC request to get the ledger's lastAddConfirm.
> For the above case, if bk1 holds the max entry 14, bk2 holds the max entry 13 and bk3 holds the max entry 14 but it is lost, the LedgerHandle get lastAddConfirm will be 13, not 14.

I have a question about this. 

```
public void checkLedger(final LedgerHandle lh,
                            final GenericCallback<Set<LedgerFragment>> cb,
                            long percentageOfLedgerFragmentToBeVerified)
```

There are two places to invoke LedgerChecker#checkLedger.

1. https://github.com/apache/bookkeeper/blob/35e9da9b55b5d44459d3421e8704be47afc6f914/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorCheckAllLedgersTask.java#L196-L212

2.https://github.com/apache/bookkeeper/blob/35e9da9b55b5d44459d3421e8704be47afc6f914/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicationWorker.java#L449-L451

Both places use the openLedgerNoRecovery to open the  LedgerHandle.

It will use lh.asyncReadLastConfirmed() to get the LAC.
https://github.com/apache/bookkeeper/blob/35e9da9b55b5d44459d3421e8704be47afc6f914/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerOpenOp.java#L222-L245

And we use the V2 protocol, it invokes `asyncReadPiggybackLastConfirmed`.
https://github.com/apache/bookkeeper/blob/35e9da9b55b5d44459d3421e8704be47afc6f914/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java#L1388-L1395

In asyncReadPiggybackLastConfirmed, it invokes ReadLastConfirmedOp#initiate() to get the LAC. 

https://github.com/apache/bookkeeper/blob/35e9da9b55b5d44459d3421e8704be47afc6f914/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java#L1422-L1428


In the ReadLastConfirmedOp, it will read all bookies in the currentEnsemble. The currentEnsemble is `15=[bk1:3181, bk2:3181, bk4:3181]`.  Not the {0=[bk1:3181, bk2:3181, bk3:3181]. 
So it will send readLac RPC to bk1, bk2, and bk4. Then get response from bk1, bk2, and bk4.
When ReadLastConfirmedOp#readEntryComplete, it will compare the response from different bookies, then pick the max lac response to override maxRecoveredData. Then invoke the callback using the maxRecoveredData.

So if the bk1 lac is 14, the ledgerHandle lac should be 14, not 13.


https://github.com/apache/bookkeeper/blob/35e9da9b55b5d44459d3421e8704be47afc6f914/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ReadLastConfirmedOp.java#L197-L154

line_l08-line_112, it will pick the max lac to override maxRecoveredData.

line_137-line_148, use the maxRecoveredData to invoke callback.

**Comment by hangc0276** [comment]
> > How the lastAddConfirm was generated
> > Due to the ledger being in the OPEN state, the ledger handle will send a readLAC request to get the ledger's lastAddConfirm.
> > For the above case, if bk1 holds the max entry 14, bk2 holds the max entry 13 and bk3 holds the max entry 14 but it is lost, the LedgerHandle get lastAddConfirm will be 13, not 14.
> 
> I have a question about this.
> 
> ```
> public void checkLedger(final LedgerHandle lh,
>                             final GenericCallback<Set<LedgerFragment>> cb,
>                             long percentageOfLedgerFragmentToBeVerified)
> ```
> 
> There are two places to invoke LedgerChecker#checkLedger.
> 
> 1. https://github.com/apache/bookkeeper/blob/35e9da9b55b5d44459d3421e8704be47afc6f914/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorCheckAllLedgersTask.java#L196-L212
> 
> 2.
> 
> https://github.com/apache/bookkeeper/blob/35e9da9b55b5d44459d3421e8704be47afc6f914/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicationWorker.java#L449-L451
> 
> Both places use the openLedgerNoRecovery to open the LedgerHandle.
> 
> It will use lh.asyncReadLastConfirmed() to get the LAC.
> 
> https://github.com/apache/bookkeeper/blob/35e9da9b55b5d44459d3421e8704be47afc6f914/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerOpenOp.java#L222-L245
> 
> And we use the V2 protocol, it invokes `asyncReadPiggybackLastConfirmed`.
> 
> https://github.com/apache/bookkeeper/blob/35e9da9b55b5d44459d3421e8704be47afc6f914/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java#L1388-L1395
> 
> In asyncReadPiggybackLastConfirmed, it invokes ReadLastConfirmedOp#initiate() to get the LAC.
> 
> https://github.com/apache/bookkeeper/blob/35e9da9b55b5d44459d3421e8704be47afc6f914/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java#L1422-L1428
> 
> In the ReadLastConfirmedOp, it will read all bookies in the currentEnsemble. The currentEnsemble is `15=[bk1:3181, bk2:3181, bk4:3181]`. Not the {0=[bk1:3181, bk2:3181, bk3:3181]. So it will send readLac RPC to bk1, bk2, and bk4. Then get response from bk1, bk2, and bk4. When ReadLastConfirmedOp#readEntryComplete, it will compare the response from different bookies, then pick the max lac response to override maxRecoveredData. Then invoke the callback using the maxRecoveredData.
> 
> So if the bk1 lac is 14, the ledgerHandle lac should be 14, not 13.
> 
> https://github.com/apache/bookkeeper/blob/35e9da9b55b5d44459d3421e8704be47afc6f914/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ReadLastConfirmedOp.java#L97-L154
> 
> line_l08-line_112, it will pick the max lac to override maxRecoveredData.
> 
> line_137-line_148, use the maxRecoveredData to invoke callback.

@horizonzy Yes, you are right. I updated the description, please help take a look, thanks.

**Comment by hangc0276** [comment]
> I wonder of we risk to lose one entry in case of AQ < WQ.
> 
> I have to think more

@eolivelli Sorry, I misunderstood the lastAddConfirm part. It doesn't have the risk of losing one entry. I updated the description, please help take a look, thanks.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragment.java [hunk 2]
```diff
@@ -53,8 +53,13 @@ public LedgerFragment(LedgerHandle lh,
...
         SortedMap<Long, ? extends List<BookieId>> ensembles = lh
                 .getLedgerMetadata().getAllEnsembles();
+        // Check if the ledger fragment is closed has two conditions
+        // 1. The ledger is closed
+        // 2. This fragment is not the last fragment and this ledger's lastAddConfirm >= ensembles.lastKey() - 1.
+        //    This case happens when the ledger's last ensemble is empty
         this.isLedgerClosed = lh.getLedgerMetadata().isClosed()
-                || !ensemble.equals(ensembles.get(ensembles.lastKey()));
+                || (!ensemble.equals(ensembles.get(ensembles.lastKey()))
+            && lh.getLastAddConfirmed() >= ensembles.lastKey() - 1);
```
Comment (horizonzy) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragment.java:62]: Here should be `lh.getLastAddConfirmed() >= ensembles.lastKey() - 2`.
The last entry in the bookie may be entry 9, the getLastAddConfirmed is 9-1 = 8. The last ensemble key maybe 10. 8 >= 10 - 2.

Comment (hangc0276) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragment.java:62]: Maybe not. Take the following case for example:
0: [bk1, bk2, bk3]
10: [bk1, bk2, bk4]
LAC = 100
Ledger is OPEN

For the first ensemble: isLedgerClosed = (false || (true && 100 >= 9)) => true
For the last ensemble: isLedgerClosed = (false || (false && 100 >= 9)) => false
The result is expected.

Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragment.java:62]: Yes, you are right.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/replication/TestReplicationWorker.java [hunk 3]
```diff
@@ -1158,6 +1158,33 @@ public void testRWShutDownInTheCaseOfZKOperationFailures() throws Exception {
         }
     }
 
+    @Test
+    public void testReplicateEmptyOpenStateLedger() throws Exception {
+        LedgerHandle lh = bkc.createLedger(3, 3, 2, BookKeeper.DigestType.CRC32, TESTPASSWD);
+        assertFalse(lh.getLedgerMetadata().isClosed());
+
+        List<BookieId> firstEnsemble = lh.getLedgerMetadata().getAllEnsembles().firstEntry().getValue();
```
Comment (horizonzy) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/replication/TestReplicationWorker.java:1166]: `firstEnsemble` is unused

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicationWorker.java [hunk 4]
```diff
@@ -561,6 +563,14 @@ private boolean rereplicate(long ledgerIdToReplicate) throws InterruptedExceptio
...
+     * <p>Another corner case is that there are multiple ensembles in the ledger and the last
+     * segment/ensemble is open, but nothing has been written to some quorums in the ensemble.
+     * For the v2 protocol, this ledger's lastAddConfirm entry is the last segment/ensemble's `key - 2`,
+     * not `key - 2`, the explanation please refer to: https://github.com/apache/bookkeeper/pull/3917.
+     * If we treat the penultimate segment/ensemble as closed state, we will can't replicate
+     * the last entry in the segment. So in this case, we should also check if the penultimate
+     * segment/ensemble has missing bookies.
      */
     private boolean isLastSegmentOpenAndMissingBookies(LedgerHandle lh) throws BKException {
         LedgerMetadata md = admin.getLedgerMetadata(lh);
```
Comment (horizonzy) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicationWorker.java:576]: `md` is unused.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicationWorker.java [hunk 5]
```diff
@@ -570,6 +580,10 @@ private boolean isLastSegmentOpenAndMissingBookies(LedgerHandle lh) throws BKExc
 
         SortedMap<Long, ? extends List<BookieId>> ensembles = admin.getLedgerMetadata(lh).getAllEnsembles();
         List<BookieId> finalEnsemble = ensembles.get(ensembles.lastKey());
+        if (ensembles.size() > 1 && lh.getLastAddConfirmed() < ensembles.lastKey() - 1) {
+            finalEnsemble = new ArrayList<>(finalEnsemble);
```
Comment (horizonzy) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicationWorker.java:584]: `finalEnsemble = new ArrayList<>(finalEnsemble);` may be meaningless.

Comment (hangc0276) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicationWorker.java:584]: The `finalEnsemble` is immutable in the ledger metadata.

Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicationWorker.java:584]: Oh, get it.

**Comment by hangc0276** [comment]
@merlimat @eolivelli @dlg99 Please help take a look, thanks.

**Review by eolivelli — APPROVED** [review]
+1 LGTM

**Review by wenbingshen — APPROVED** [review]
LGTM

**Pull request merged**